Display Tab Indexes in VSCode

swordjjjkkk opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->
I'd like a setting to decide show tab index.
there are two reasons i need to open this pull request.

1.It's useful for vim extension use \<leader\>+num to switch tab pages , and this won't disturb regular users, as it is disabled by default.
2.I tried compile an oss version , but the remote-ssh Extension not support oss version.

I look forward to any feedback or suggestions for improvement. Thank you for considering.

swordjjjkkk commented:
@microsoft-github-policy-service agree

moataz2002 commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -56,6 +56,7 @@ import { IEditorTitleControlDimensions } from 'vs/workbench/browser/parts/editor
 import { StickyEditorGroupModel, UnstickyEditorGroupModel } from 'vs/workbench/common/editor/filteredEditorGroupModel';
 import { IReadonlyEditorGroupModel } from 'vs/workbench/common/editor/editorGroupModel';
```
Important Step to be done is the import part

moataz2002 commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -149,6 +150,7 @@ export class MultiEditorTabsControl extends EditorTabsControl {
 		@IPathService private readonly pathService: IPathService,
 		@ITreeViewsDnDService private readonly treeViewsDragAndDropService: ITreeViewsDnDService,
```
edited the configuration is made

moataz2002 commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -1458,7 +1465,14 @@ export class MultiEditorTabsControl extends EditorTabsControl {
+
+			const enabled = this.configurationService.getValue<boolean>('workbench.tabIndex.enabled');
+
+			if (enabled) {
```
tabIndex. is enabled and making the desired condition

moataz2002 commented on src/vs/workbench/contrib/preferences/browser/settingsLayout.ts:
```diff
@@ -101,7 +101,7 @@ export const tocData: ITOCEntry<string> = {
 					id: 'workbench/appearance',
 					label: localize('appearance', "Appearance"),
-					settings: ['workbench.activityBar.*', 'workbench.*color*', 'workbench.fontAliasing', 'workbench.iconTheme', 'workbench.sidebar.location', 'workbench.*.visible', 'workbench.tips.enabled', 'workbench.tree.*', 'workbench.view.*']
+					settings: ['workbench.activityBar.*', 'workbench.*color*', 'workbench.fontAliasing', 'workbench.iconTheme', 'workbench.sidebar.location', 'workbench.*.visible', 'workbench.tips.enabled', 'workbench.tabIndex.enabled', 'workbench.tree.*', 'workbench.view.*']
```
edited settings handled for the new modifications

benibenj replied:
There should not be a change here

moataz2002 reviewed (approved):
Nice to be merged

benibenj commented:
Please create an issue for this feature and link this PR. Make sure to add a detailed description which describes the use case and the problem it solves.

FYI we added custom editor labels in this months release. It allows you change the name of the open editors/tabs. The setting name is `workbench.editor.customLabels.patterns`. But it won't give you any info on the tab index.

swordjjjkkk commented:
> Please create an issue for this feature and link this PR. Make sure to add a detailed description which describes the use case and the problem it solves.
> 
> FYI we added custom editor labels in this months release. It allows you change the name of the open editors/tabs. The setting name is `workbench.editor.customLabels.patterns`. But it won't give you any info on the tab index.

fixes #209533

bpasero commented:
> FYI we added custom editor labels in this months release. It allows you change the name of the open editors/tabs. The setting name is workbench.editor.customLabels.patterns. But it won't give you any info on the tab index.

I like the idea of having a new variable for the tab index for custom editor labels. Maybe one that is group-relative and one that is global?

ghrejoe commented:
> Please create an issue for this feature and link this PR. Make sure to add a detailed description which describes the use case and the problem it solves.
> 
> FYI we added custom editor labels in this months release. It allows you change the name of the open editors/tabs. The setting name is `workbench.editor.customLabels.patterns`. But it won't give you any info on the tab index.

It would be great if we could add it to this setting

benibenj commented:
I would disagree with making this part of the `workbench.editor.customLabels.patterns` setting. Custom Labels have no relation to tabs/editor groups state. I would rather add a new setting `workbench.editor.showIndex` or something like this. This would just prepend the index of the editor in the group to the tab. I also check if we could make it part of one of the label format settings but I don't think that would work. What do you think @ghrejoe @bpasero?

benibenj commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -160,6 +162,11 @@ export class MultiEditorTabsControl extends EditorTabsControl {
 
 		// React to decorations changing for our resource labels
 		this._register(this.tabResourceLabels.onDidChangeDecorations(() => this.doHandleDecorationsChange()));
+		this._register(this.configurationService.onDidChangeConfiguration(e => {
```
We have `updateOptions` which is called when the options change. You can update the editor label there if you see this settings changed

benibenj commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -1458,7 +1465,14 @@ export class MultiEditorTabsControl extends EditorTabsControl {
+			const enabled = this.configurationService.getValue<boolean>('workbench.tabIndex.enabled');
+
+			if (enabled) {
+				name = (tabIndex + 1).toString() + '. ' + tabLabel.name;
```
nit `name = showIndex ? '${tabIndex+1}. ${tabLabel.name}' : tabLabel.name`

benibenj commented on src/vs/workbench/browser/workbench.contribution.ts:
```diff
@@ -619,6 +619,11 @@ const registry = Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Con
 				'default': true,
 				'description': localize('tips.enabled', "When enabled, will show the watermark tips when no editor is open.")
 			},
+			'workbench.tabIndex.enabled': {
```
setting name should align with other setting names. Something like `workbench.editor.showTabIndex`

benibenj commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -1458,7 +1465,14 @@ export class MultiEditorTabsControl extends EditorTabsControl {
 		} else {
-			name = tabLabel.name;
+
+			const enabled = this.configurationService.getValue<boolean>('workbench.tabIndex.enabled');
```
setting should be available in `options`

benibenj reviewed (changes requested):
Just a quick review. Haven't tested yet. Also there is a merge conflict. But we still need to finish discussing if and how this should exactly work.

swordjjjkkk committed:
rebase

swordjjjkkk committed:
1.Moved the settings into options.
2.Moved tab refresh to updateOptions.
3.Simplified the code using a ternary operator.
4.Renamed the setting from 'workbench.tabIndex.enabled' to 'workbench.editor.showTabIndex'.
5.Removed some irrelevant code.

bpasero commented on src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts:
```diff
@@ -1600,7 +1601,8 @@ export class MultiEditorTabsControl extends EditorTabsControl {
 		} else {
-			name = tabLabel.name;
+
+			name = options.showTabIndex ? `${tabIndex + 1}. ${tabLabel.name}` : tabLabel.name
```
This is missing `nls.localize` call so that it can be properly localised for all the languages we support.

swordjjjkkk replied:
The main purpose of this feature is to display the position of the tab, so that it is convenient to use keyboard shortcuts to jump to the corresponding tab. Therefore, hard coding is not acceptable. You can refer to this issue for more detailed information.#209533

swordjjjkkk replied:
I might have misunderstood what you said. Could you please tell me exactly what to do?

swordjjjkkk replied:
Sorry, I am still confused. The original code you referenced is name = tabLabel.name;. I just added a condition to check the setting and then prefixed it with a number. The number does not need nls.localize, right? The tabLabel.name is the filename, which also does not require nls.localize, right? So where exactly should the change be made, or did you quote the wrong part of the code?

bpasero replied:
I think even when you have 2 dynamic parts separated by `.` and space you might have a language where `1. something` is not the standard form (I dont know?). I typically stay away from making the decision across all languages by ensuring to localize whenever I construct a string that is presented in the UI somehow.

bpasero replied:
On the other hand, now that I thought more about it, it seems that you do not see this as something that follows normal language style where something is enumerated. You see this just as a way to present the index in front of the file label?

If that is the case, I wonder if it should really be made more distinct from the file name so that people are not confusing it with the actual file name 🤔

swordjjjkkk replied:
Got it, I understand. Thank you for your explanation, I will make the changes as soon as possible.

swordjjjkkk replied:
I have added the localization handling. If there is anything else that needs to be modified, please let me know.

swordjjjkkk commented:
I have made the code modifications as required, specifically modifying the following:
1.Moved the settings into options.
2.Moved tab refresh to updateOptions.
3.Simplified the code using a ternary operator.
4.Renamed the setting from 'workbench.tabIndex.enabled' to 'workbench.editor.showTabIndex'.
5.Removed some irrelevant code.

bpasero reviewed (dismissed):
I am still not 100% sold that this cannot be configured through editor custom labels. We now limit this feature to how the index appears by hardcoding it (using just the number or number + dot) when people maybe want to configure this (see https://github.com/mkitt/tabline.vim where colon is used).

swordjjjkkk committed:
1. add localize  for tabindex display 2. change the default format to use a colon

roachsinai commented:
Any updates?

ghrejoe commented:
Excuse me, how is it going now

swordjjjkkk commented:
> Excuse me, how is it going now
The code is already finished, and I have also modified it according to the requirements. Then there was no further message, and I don't know what to do next either.

swordjjjkkk commented:
> Excuse me, how is it going now

The code is already finished, and I have also modified it according to the requirements. Then there was no further message, and I don't know what to do next either.

benibenj committed:
Merge branch 'main' into tabindex

benibenj commented:
@bpasero Custom Labels work solely based on the file path. It's possible that a custom label is rendered for a file which is not opened. In that case no index could be resolved. 

Imagine the template `${tabIndex}: ${filename}.${extname}`. If we render the custom label for a file which is not open for example in quick open `${tabIndex}``could not be resolved to a number, hence we would have to render an empty string for it and so it would display `: file.txt` which is not wanted. 

Also, if custom labels would use editor group information, it would probably strongly impact performance.

I wonder if we could still allow this to be customized  without using custom labels

benibenj committed:
semicolumn missing

benibenj committed:
Merge branch 'main' into tabindex

vivodi commented:
Is this still relevant?

benibenj committed:
Merge branch 'main' into tabindex

benibenj committed:
fix tab index for untitled files

bpasero reviewed (changes requested):
I will review later

bpasero reviewed (dismissed):
This is not in a good state, esp. around the last commit that adds knowledge about how to render untitled into the tabs control.

I would expose a new option `prefix` on the label that is passed down based on this setting.

benibenj committed:
add namePrefix option for resource labels

bpasero committed:
Merge branch 'main' into tabindex

bpasero reviewed (approved):
The display of tab index cannot be localised, which I think is fair considering we want to define how the experience is, but just wanted to note that maybe not in all languages `{number}: {name}` is natural.

roachsinai commented:
Thanks a lot!

ghrejoe commented:
I've already experienced this feature in the Code Insiders version, and I find it very useful.
